src/audio/component.c

```diff
@@ -506,6 +516,54 @@ int comp_copy(struct comp_dev *dev)
 	return ret;
 }
 
+#ifdef CONFIG_SOF_TELEMETRY_PERFORMANCE_MEASUREMENTS
+void comp_init_performance_data(struct comp_dev *dev)
+{
+	struct perf_data_item_comp *item = dev->perf_data.perf_data_item;
+
+	if (item)
+		perf_data_item_comp_init(item, dev->ipc_config.id, 0);
+}
+
+/* returns true if budget violation occurred */
+static bool update_peak_of_measured_cpc(struct comp_dev *dev, size_t measured_cpc)
+{
+	if (measured_cpc <= dev->perf_data.peak_of_measured_cpc)
+		return false;
+	dev->perf_data.peak_of_measured_cpc = measured_cpc;
+	return measured_cpc > dev->cpc;
+}
+
+bool comp_update_performance_data(struct comp_dev *dev, uint32_t cycles_used)
+{
+	struct perf_data_item_comp *item = dev->perf_data.perf_data_item;
+
+	if (perf_meas_get_state() == IPC4_PERF_MEASUREMENTS_STARTED) {
+		/* we divide by ibs so we need to check if its set */
+		if (item && dev->ibs != 0) {
```

Comment:
in `comp_update_ibs_obs_cpc()` in an error case you set `ibs` to 1, which presumably doesn't produce correct performance results. Maybe it's netter to set it to 0 there to skip all the calculations here?

Reply:
Good idea.